server/api_archive.go

```diff
@@ -0,0 +1,170 @@
+package server
+
+// This is a low-level archiver which zips and sends GCA server persistent file
+// data using a POST command. This is intended to be used by an external service
+// to provide archival snapshots of raw data in the system. Only public data is
+// provided, with private keys zeroed before sending.
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+)
+
+// ArchiveHandler provides the POST /archive api endpoint. It returns
+// uninterpreted file data intended to be used by an external service
+// to produce an archive.
+func (gcas *GCAServer) ArchiveHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		http.Error(w, "Only GET method is allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	if r.ContentLength != 0 {
+		http.Error(w, "Request should not have a body", http.StatusBadRequest)
+		return
+	}
+
+	if !gcas.ApiArchiveRateLimiter.Allow() {
+		http.Error(w, fmt.Sprintf("Too many requests, this server allows %v every %v", apiArchiveLimit, apiArchiveRate), http.StatusTooManyRequests)
+		return
+	}
+
+	buffer := new(bytes.Buffer)
+	znw := zip.NewWriter(buffer)
+
+	// File ordering matters here. We are not locking the files, and relying on the file writes being
+	// append only, and happening in a single write. We must also archive files later which depend
+	// on earlier files.
+
+	// Add the pseudo file server.pubkey
+	if err := gcas.addPubKeyFile("server.pubkey", znw); err != nil {
+		http.Error(w, fmt.Sprintf("Error zipping server.pubkey: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	for _, name := range PublicFiles() {
+		err := gcas.addFile(name, znw)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Error zipping file %v: %v", name, err), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	// Add a README
+	if err := gcas.addReadmeFile(znw); err != nil {
+		http.Error(w, fmt.Sprintf("Error adding README: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	if err := znw.Close(); err != nil {
+		if _, err := w.Write(buffer.Bytes()); err != nil {
+			http.Error(w, "Failed to close zip writer", http.StatusInternalServerError)
+			return
+		}
+	}
+
+	w.Header().Set("Content-Type", "application/zip")
+
+	if _, err := w.Write(buffer.Bytes()); err != nil {
+		http.Error(w, "Failed to write archive response", http.StatusInternalServerError)
+	}
```

Comment:
have you manually tested this endpoint in a browser to confirm that it returns the expected data?

Reply:
```
$ curl -o archive.zip localhost:35015/api/v1/archive
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  4257    0  4257    0     0   222k      0 --:--:-- --:--:-- --:--:--  230k
$ unzip -t  archive.zip 
Archive:  archive.zip
    testing: allDeviceStats.dat       OK
    testing: equipment-reports.dat    OK
    testing: equipment-authorizations.dat   OK
    testing: gcaPubKey.dat            OK
    testing: gcaTempPubKey.dat        OK
    testing: server.pubkey            OK
    testing: README                   OK
No errors detected in compressed data of archive.zip.
```